api/buckets/service.go

```diff
@@ -1166,24 +1203,71 @@ func getLink(lnks []*ipld.Link, name string) *ipld.Link {
 // updateOrAddPin moves the pin at from to to.
 // If from is nil, a new pin as placed at to.
 func (s *Service) updateOrAddPin(ctx context.Context, from, to path.Path) error {
```

Comment:
So a bit of a refactor in `updateOrAddPin`.
Basically, it does the same as it did before (with some if-nesting improvement).
Additionally, it does:
- Get a cached total size of buckets attribute at the Account level.
- Get size of the root which will be unpined (or 0 if there isn't)
- Get size of the root to be pinned.
- Calculate the "future" size of this bucket (before moving any pin)
- Check the limit. If not ok, fail. (No pin was moved)
- If ok, move on and update the total size of the account.

Quite an important method since is used in anything that updates the bucket root, so might worth extra review attention just in case.

Reply:
Just combing though it all myself to see if all cases are handled.
- What about deleting a bucket? There should probably be some handling in `Remove`.
- Most likely this is not handling the encrypted dags properly, where cumulative size is unable to walk the encrypted links. I think we'd need some custom version that uses `getNodesToPath`. 
- There are some cases where nodes are unpinned when handling encrypted dags, i.e., `unpinNodeAndBranch` is used in `SetPath` and indirectly in `RemovePath`
- There are some cases where nodes are pinned directly outside of `updateOrAddPin`, i.e., in `insertNodeAtPath` and in `SetPath` where the encryption key is not nil and we're not replacing the entire root path (`addAndPinNodes`)
- `createBootstrappedPath` could lead to an explosion

That's from a first pass, there may be other edge cases. However, handling all these directly sounds like a bit of a nightmare. So, for these bucket sizes, maybe we can only consider actual files (`RawNode` or `ProtoNode`s that wrap `RawNodes`, which are larger files), and ignore tracking the sizes of _all_ directory nodes, encrypted or not. I think that would just mean tracking a size _in the Bucket schema_, (which might be nice to have anyway), and using that number to reject new incoming files. Thoughts?